openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py

```diff
@@ -0,0 +1,210 @@
+"""
+ObjectTags for System-defined Taxonomies
+"""
+from enum import Enum
+from typing import List
+
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from django.core.exceptions import FieldDoesNotExist
+from django.db import models
+
+
+from openedx_tagging.core.tagging.models import (
+    Taxonomy,
+    OpenObjectTag,
+    ClosedObjectTag,
+)
+from openedx_tagging.core.tagging.registry import register_object_tag_class
+
+
+class SystemDefinedIds(Enum):
+    """
+    System-defined taxonomy IDs
+    """
+    LanguageTaxonomy = 1
+
+
+class SystemDefinedObjectTagMixin:
+    """
+    Mixing for ObjectTags used on all system defined taxonomies
+
+    `system_defined_taxonomy_id``is used to connect the 
+    ObjectTag with the system defined taxonomy.
+    This is because there can be several ObjectTags
+    for the same Taxonomy, ex:
+    
+    - LanguageCourseObjectTag
+    - LanguageBlockObjectTag
+
+    On the example, there are ObjectTags for the same Language
+    taxonomy but with different objects.
+
+    Using this approach makes the connection between the ObjectTag
+    and system defined taxonomy as hardcoded and can't be changed.
+    """
+
+    system_defined_taxonomy_id = None
+
+    def _check_system_taxonomy(self, taxonomy: Taxonomy = None):
+        """
+        Validates if the taxonomy is system-defined and match
+        with the name stored in the object tag
+        """
+        return (
+            bool(taxonomy) and
+            taxonomy.system_defined and
+            taxonomy.id == self.system_defined_taxonomy_id
+        )
+
+
+class OpenSystemObjectTag(OpenObjectTag, SystemDefinedObjectTagMixin):
+    """
+    Free-text object tag used on system-defined taxonomies
+    """
+
+    class Meta:
+        proxy = True
+
+    def _check_taxonomy(self):
+        return (
+            super()._check_taxonomy() and
+            self._check_system_taxonomy(self.taxonomy)
+        )
+
+    
+class ClosedSystemObjectTag(ClosedObjectTag, SystemDefinedObjectTagMixin):
+    """
+    Object tags linked to a closed system-taxonomy
+    """
+
+    class Meta:
+        proxy = True
+
+    def _check_taxonomy(self):
+        return (
+            super()._check_taxonomy() and
+            self._check_system_taxonomy(self.taxonomy)
+        )
+
+
+class ModelObjectTag(OpenSystemObjectTag):
+    """
+    Object tags used with tags that relate to the id of a model
+    
+    This object tag class is not registered as it needs to have an associated model
+    """
+
+    class Meta:
+        proxy = True
+
+    tag_class_model = None
+
+
+    def _check_taxonomy(self):
+        """
+        Validates if has an associated Django model that has an Id
+        """
+        if not super()._check_taxonomy():
+            return False
+    
+        if not self.tag_class_model:
+            return False
```

Comment:
Similarly here, if we use an "unimplemented" instance property method instead of a class constant, subclasses will be forced to implement it, or face a compile-time error.

Also, because class-level inheritance for python classes isn't so great, using an instance property allows sub-subclasses to override this property as well (if anyone ever wants to do that).

```suggestion

    @property
    def tag_class_model(self) -> Type:
        """
        Subclasses must implement this method to return the Django.model
        class referenced by these object tags.
        """
        raise NotImplementedError

    def _check_taxonomy(self):
        """
        Validates if has an associated Django model that has an Id
        """
        if not super()._check_taxonomy():
            return False
```